### ISSUE TITLE
Swagger import support?

### ISSUE THREAD

**Issue by lyptt** [lonelycode/tyk#65]
Are there any plans to support Swagger import? I've noticed there's API blueprint import, however we export our documentation via code annotations, which doesn't seem to have the capability to do so in API Blueprint format.

From looking at the blueprint.go file, I can see the Blueprint import seems to be quite simple, however it's also fairly specific to that. I'm not entirely sure how to go about building in support for Swagger myself, so if there's no current plans for Swagger import, advice in this area would be helpful.

**Comment by lonelycode** [comment]
There are plans to eventually support it, though it's not on the roadmap atm.

The importer is actually an interface, it's very easy to extend the functionality simply by [implementing the importer interface](https://github.com/lonelycode/tyk/blob/master/blueprint.go#L16) and registering it as a name.

The way the importers work is quite simple really - you need to create a `struct` that resembles the swagger format (or can capture it), and the transform it into our own `ApiDefinition` format (That's what those private methods are for)

All the bits are there for generalising to a multi-format solution, they just haven't been tied together :-/

The trickier bit would be to generalise the [command mode options](https://github.com/lonelycode/tyk/blob/master/command_mode.go) - as they aren't as modular as they could be. Though you could jsut create a new specific flag handler and then invoke your converter - the longer term plan wa to generalise the whole thing and make it so that importers could jut be "registered".

Bottom line - the whole thing needs a little polish :-S

**Comment by lyptt** [comment]
Okay great, I've started working on the Swagger import. I'm assuming there won't be automation support for it until that's been worked on by you? I haven't found any references in the source code to API Blueprint import other than command mode, so I'm assuming that's all in the Dashboard somewhere.

**Comment by lonelycode** [comment]
Hi There,

Yes all of that lives in the dashboard, it's actually shared code - we just need to pull it in. Also, the command mode option will only dump the APi definition to file / memory, not into the DB. 

But if you put the basics in, we'll roll it up :-)

Thanks,
Martin

**Comment by lonelycode** [comment]
Thought id add this here:

http://mholt.github.io/json-to-go/

Really handy for generating tha massive Swagger struct, if you can get hold of a fully qualified AST for Swagger this will make writing it less painful ;-)

**Comment by lyptt** [comment]
That'll be handy, thanks! We'll probably be taking this on properly in a couple of days time.

**Comment by mingfang** [comment]
+1

**Comment by grocky** [comment]
:+1:

**Comment by lonelycode** [comment]
Fixed in master, can now be accessed using --import-swagger command line option (same as blueprint). dashboard interface TBC, on roadmap for 1.7.

**Comment by mingfang** [comment]
Thank you very much!

**Comment by lonelycode** [comment]
Only basic supprot :-) Mocks not supported (Swagger doesn;t seem to let you define mock responses very well, just schemas)

**Comment by smuniswamy** [comment]
@lonelycode, I noticed latest Tyk Dashboard docker image includes support for swagger import. Is there an ETA on v0.9.5.4 release?

**Comment by lonelycode** [comment]
Hi,

The current dashboard release that is in our GitHub releases section is the same as the one in the docker image, which includes swagger support on imports. So you should have that feature available now.

The next version of the dashboard will be out in the next few months - we have quite a bit to do - take a look at our roadmap:

https://trello.com/b/59d5kAZ5/tyk-api-gateway-roadmap

Cheers,
Martin

**Comment by smuniswamy** [comment]
Thanks @lonelycode. I have upgraded from v0.9.3 to v0.9.5.3 but did not see Swagger option. Just found out it was due to browser cache.
